src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleDictionary.cs

```diff
@@ -0,0 +1,92 @@
+        {
+            Debug.Assert(!state.Current.IsProcessingEnumerable);
+
+            JsonPropertyInfo jsonPropertyInfo = state.Current.JsonPropertyInfo;
+            if (jsonPropertyInfo == null)
+            {
+                jsonPropertyInfo = state.Current.JsonClassInfo.CreateRootObject(options);
+            }
+
+            Debug.Assert(jsonPropertyInfo != null);
+
+            // A nested object or dictionary so push new frame.
+            if (state.Current.PropertyInitialized)
+            {
+                Debug.Assert(state.Current.IsDictionary);
+
+                JsonClassInfo classInfoTemp = state.Current.JsonClassInfo;
+                state.Push();
+                state.Current.JsonClassInfo = classInfoTemp.ElementClassInfo;
+                state.Current.InitializeJsonPropertyInfo();
+
+                ClassType classType = state.Current.JsonClassInfo.ClassType;
+                if (classType == ClassType.Value &&
+                    jsonPropertyInfo.ElementClassInfo.Type != typeof(object) &&
+                    jsonPropertyInfo.ElementClassInfo.Type != typeof(JsonElement))
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.JsonClassInfo.Type, reader, state.PropertyPath);
+                }
+
+                JsonClassInfo classInfo = state.Current.JsonClassInfo;
+                state.Current.ReturnValue = classInfo.CreateObject();
+                return;
+            }
+
+            state.Current.PropertyInitialized = true;
+
+            // If current property is already set (from a constructor, for example) leave as-is.
+            if (jsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue) == null)
+            {
+                // Create the dictionary.
+                JsonClassInfo dictionaryClassInfo = options.GetOrAddClass(jsonPropertyInfo.RuntimePropertyType);
+                IDictionary value = (IDictionary)dictionaryClassInfo.CreateObject();
+                if (value != null)
+                {
+                    if (state.Current.ReturnValue != null)
+                    {
+                        state.Current.JsonPropertyInfo.SetValueAsObject(state.Current.ReturnValue, value);
+                    }
+                    else
+                    {
+                        // A dictionary is being returned directly, or a nested dictionary.
+                        state.Current.SetReturnValue(value);
+                    }
+                }
+            }
+        }
+
+        private static void HandleEndDictionary(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack state)
+        {
+            if (state.Current.IsDictionaryProperty)
+            {
+                // We added the items to the dictionary already.
+                state.Current.ResetProperty();
+                return;
+            }
+
+            object value = state.Current.ReturnValue;
+
+            if (state.IsLastFrame)
```

Comment:
Can you add a comment explaining why we don't `ApplyObjectToEnumerable` in this case?